### ISSUE TITLE
Cannot set proxy_config tag in IoT Edge API Proxy.  Limit exceeds the 4096 tag limit

### ISSUE THREAD

**Issue by ksaye** [Azure/iotedge#5269]
## Expected Behavior
Following the instructions at https://docs.microsoft.com/en-us/azure/iot-edge/how-to-configure-api-proxy-module?view=iotedge-2020-11#edit-the-proxy-configuration, I wanted to modify the nginx_default_config.conf.

The instructions say to take the template and modify it, then base64 encode it and set it as the value proxy_config in the IoT Edge API Proxy twin.

When I take the template, no changes, and base64 encode it, the character count is 7012, which is greater than the 4096 tag limit.

Even if I remove some un needed items, I can only get it to 5000 characters.

## Current Behavior
I get the error:

```
An error occurred while submitting.  The request contained incorrectly formatted information.

Details: ErrorCode:ArgumentInvalid;Property or Tag value should be maximum 4096 bytes. Error in Property/Tag ZXZlbnRzIHsgfQoKaHR0cCB7CiBwcm94eV9idWZmZXJzIDMyIDE2MGs7ICAKIHByb3h5X2J1ZmZlcl9zaXplIDE2MGs7CiBwcm94eV9yZWFkX3RpbWVvdXQgMzYwMDsKIGVycm9yX2xvZyAvZGV2L3N0ZG91dCBpbmZvOwogYWNjZXNzX2xvZyAvZGV2L3N0ZG91dDsKCiBtYXAgJGh0dHBfYXV0aG9yaXphdGlvbiAkYXV0aCB7CiAgZGVmYXVsdCAgJGh0dHBfYXV0aG9yaXphdGlvbjsKICAiIiAkdG9rZW47CiB9CgogbWFwICIiICRlZGdlX2h1Yl9lbmRwb2ludCB7CiAgZGVmYXVsdCAgaHR0cHM6Ly9lZGdlaHViOwogICIiICBodHRwczovL2VkZ2VodWI7CiB9CgogcHJveHlfY2FjaGVfcGF0aCAvdmFyL2NhY2hlL25naW54IGxldmVscz0xIGtleXNfem9uZT10b2tlbl9jYWNoZToxbSBtYXhfc2l6ZT0ybQogIGluYWN0aXZlPTYwbSB1c2VfdGVtcF9wYXRoPW9mZjsKCiBzZXJ2ZXIgewogbGlzdGVuICR7TkdJTlhfREVGQVVMVF9QT1JUfSBzc2wgZGVmYXVsdF9zZXJ2ZXI7CgogY2h1bmtlZF90cmFuc2Zlcl9lbmNvZGluZyBvbjsKCiBzc2xfY2VydGlmaWNhdGUgc2VydmVyLmNydDsKIHNzbF9jZXJ0aWZpY2F0ZV9rZXkgcHJpdmF0ZV9rZXlfc2VydmVyLnBlbTsgCiBzc2xfY2xpZW50X2NlcnRpZmljYXRlIHRydXN0ZWRDQS5jcnQ7CiBzc2xfdmVyaWZ5X2RlcHRoIDc7CiBzc2xfdmVyaWZ5X2NsaWVudCBvcHRpb25hbF9ub19jYTsKCiAjaWZfdGFnICR7QkxPQl9VUExPQURfUk9VVEVfQUREUkVTU30KIGlmICgkaHR0cF94X21zX3ZlcnNpb24pCiB7CiAgcmV3cml0ZSBeKC4qKSQgL3N0b3JhZ2UkMSBsYXN0OwogfQogI2VuZGlmX3RhZyAke0JMT0JfVVBMT0FEX1JPVVRFX0FERFJFU1N9IAogI2lmX3RhZyBib29sZWFuX2V4cHJlc3Npb25bISgke0JMT0JfVVBMT0FEX1JPVVRFX0FERFJFU1N9KV0KIGlmICgkaHR0cF94X21zX3ZlcnNpb24pCiB7CiAgcmV3cml0ZSBeKC4qKSQgL3BhcmVudCQxIGxhc3Q7CiB9CiAjZW5kaWZfdGFnIGJvb2xlYW5fZXhwcmVzc2lvblshKCR7QkxPQl9VUExPQURfUk9VVEVfQUREUkVTU30pXSAgCgogI2lmX3RhZyAke0RPQ0tFUl9SRVFVRVNUX1JPVVRFX0FERFJFU1N9CiBpZiAoJHVyaSB+IF4vdjIqLikKIHsKICByZXdyaXRlIF4oLiopJCAvcmVnaXN0cnkkMSBsYXN0OwogfQogI2VuZGlmX3RhZyAke0RPQ0tFUl9SRVFVRVNUX1JPVVRFX0FERFJFU1N9CgogI2lmX3RhZyBib29sZWFuX2V4cHJlc3Npb25bJighKCR7RE9DS0VSX1JFUVVFU1RfUk9VVEVfQUREUkVTU30pLCEoJHtDT05ORUNURURfQUNSX1JPVVRFX0FERFJFU1N9KSldCiBpZiAoJHVyaSB+IF4vdjIqLikKIHsKICByZXdyaXRlIF4oLiopJCAvcGFyZW50JDEgbGFzdDsKIH0KICNlbmRpZl90YWcgYm9vbGVhbl9leHByZXNzaW9uWyYoISgke0RPQ0tFUl9SRVFVRVNUX1JPVVRFX0FERFJFU1N9KSwhKCR7Q09OTkVDVEVEX0FDUl9ST1VURV9BRERSRVNTfSkpXSAKCiAjaWZfdGFnICR7RE9DS0VSX1JFUVVFU1RfUk9VVEVfQUREUkVTU30KIGxvY2F0aW9uIH5eL3JlZ2lzdHJ5LyguKikgewogIHByb3h5X2h0dHBfdmVyc2lvbiAxLjE7CiAgcmVzb2x2ZXIgMTI3LjAuMC4xMTsKICBzZXQgJGVuZHBvaW50IGh0dHA6Ly8ke0RPQ0tFUl9SRVFVRVNUX1JPVVRFX0FERFJFU1N9LyQxJGlzX2FyZ3MkYXJnczsKICBwcm94eV9wYXNzICRlbmRwb2ludDsKIH0KICNlbmRpZl90YWcgJHtET0NLRVJfUkVRVUVTVF9ST1VURV9BRERSRVNTfQoKICNpZl90YWcgJHtCTE9CX1VQTE9BRF9ST1VURV9BRERSRVNTfQogbG9jYXRpb24gfl4vc3RvcmFnZS8oLiopewogIHJlc29sdmVyIDEyNy4wLjAuMTE7CiAgcHJveHlfaHR0cF92ZXJzaW9uIDEuMTsKICBzZXQgJGVuZHBvaW50IGh0dHA6Ly8ke0JMT0JfVVBMT0FEX1JPVVRFX0FERFJFU1N9LyQxJGlzX2FyZ3MkYXJnczsKICBwcm94eV9wYXNzICRlbmRwb2ludDsKIH0KICNlbmRpZl90YWcgJHtCTE9CX1VQTE9BRF9ST1VURV9BRERSRVNTfSAKCiAjaWZfdGFnICR7SU9URURHRV9QQVJFTlRIT1NUTkFNRX0gCiBsb2NhdGlvbiB+Xi9wYXJlbnQvKC4qKSB7CiAgcHJveHlfaHR0cF92ZXJzaW9uIDEuMTsKICByZXNvbHZlciAxMjcuMC4wLjExOwogICNwcm94eV9zc2xfY2VydGlmaWNhdGUgIGlkZW50aXR5LmNydDsKICAjcHJveHlfc3NsX2NlcnRpZmljYXRlX2tleSBwcml2YXRlX2tleV9pZGVudGl0eS5wZW07CiAgcHJveHlfc3NsX3NlcnZlcl9uYW1lIG9uOwogIHByb3h5X3NzbF9uYW1lICAke0lPVEVER0VfTU9EVUxFSUR9OwogIHByb3h5X3NzbF90cnVzdGVkX2NlcnRpZmljYXRlIHRydXN0ZWRDQS5jcnQ7CiAgcHJveHlfc3NsX3ZlcmlmeV9kZXB0aCA3OwogIHByb3h5X3NzbF92ZXJpZnkgIG9uOwogIHByb3h5X3Bhc3MgIGh0dHBzOi8vJHtJT1RFREdFX1BBUkVOVEhPU1ROQU1FfToke05HSU5YX0RFRkFVTFRfUE9SVH0vJDEkaXNfYXJncyRhcmdzOwogfQogI2VuZGlmX3RhZyAke0lPVEVER0VfUEFSRU5USE9TVE5BTUV9CgogbG9jYXRpb24gfl4vZGV2aWNlc3x0d2lucy8gewogIGF1dGhfcmVxdWVzdCAvYXV0aDsKICBhdXRoX3JlcXVlc3Rfc2V0ICR0b2tlbiAkdXBzdHJlYW1faHR0cF94X3Rva2VuOwogIHJlc29sdmVyIDEyNy4wLjAuMTE7CiAgcHJveHlfaHR0cF92ZXJzaW9uICAxLjE7CiAgcHJveHlfc3NsX3ZlcmlmeSBvZmY7CiAgcHJveHlfc2V0X2hlYWRlciB4LW1zLWVkZ2UtY2xpZW50Y2VydCAkc3NsX2NsaWVudF9lc2NhcGVkX2NlcnQ7IAogIHByb3h5X3NldF9oZWFkZXIgQXV0aG9yaXphdGlvbiAkYXV0aDsgIAogIHByb3h5X3Bhc3MgJGVkZ2VfaHViX2VuZHBvaW50OwogfQoKIGxvY2F0aW9uIH5eLy5pb3RodWIvd2Vic29ja2V0IHsKICBhdXRoX3JlcXVlc3QgL2F1dGg7CiAgYXV0aF9yZXF1ZXN0X3NldCAkdG9rZW4gJHVwc3RyZWFtX2h0dHBfeF90b2tlbjsKICByZXNvbHZlciAxMjcuMC4wLjExOwogIHByb3h5X3NzbF92ZXJpZnkgb2ZmOwogIHByb3h5X3NldF9oZWFkZXIgeC1tcy1lZGdlLWNsaWVudGNlcnQgJHNzbF9jbGllbnRfZXNjYXBlZF9jZXJ0OwogIHByb3h5X3NldF9oZWFkZXIgQXV0aG9yaXphdGlvbiAkYXV0aDsgCiAgcHJveHlfcGFzcyAkZWRnZV9odWJfZW5kcG9pbnQ7CiAgcHJveHlfaHR0cF92ZXJzaW9uIDEuMTsKICBwcm94eV9zZXRfaGVhZGVyIFVwZ3JhZGUgJGh0dHBfdXBncmFkZTsKICBwcm94eV9zZXRfaGVhZGVyIENvbm5lY3Rpb24gIlVwZ3JhZGUiOwogIHByb3h5X3NldF9oZWFkZXIgSG9zdCAkaG9zdDsKIH0KIGxvY2F0aW9uIH5eLy5lc3Rwcm94eS8oLiopIHsKICByZXNvbHZlciAxMjcuMC4wLjExOwogIHByb3h5X2h0dHBfdmVyc2lvbiAgMS4xOwogIHByb3h5X3NzbF92ZXJpZnkgb2ZmOwogIHNldCAkZXN0ICR7RVNUX1VSTH0vJDEkaXNfYXJncyRhcmdzOwogIHByb3h5X3NldF9oZWFkZXIgIFgtRm9yd2FyZGVkLUhvc3QgJGh0dHBfaG9zdDsKICBwcm94eV9wYXNzICRlc3Q7CiB9CiBsb2NhdGlvbiA9IC9hdXRoIHsKICBpbnRlcm5hbDsKICBwcm94eV9jYWNoZSB0b2tlbl9jYWNoZTsKICBwcm94eV9jYWNoZV92YWxpZCBhbnkgJHtUT0tFTl9WQUxJRElUWV9NSU5VVEVTfW07CiAgcHJveHlfY2FjaGVfa2V5ICAidG9rZW4iOyAgCiAgcHJveHlfaHR0cF92ZXJzaW9uIDEuMTsKICByZXNvbHZlciAxMjcuMC4wLjExOwogIHNldCAkZW5kcG9pbnQgaHR0cDovLzEyNy4wLjAuMToke1RPS0VOX1NFUlZFUl9QT1JUfTsKICBwcm94eV9wYXNzICRlbmRwb2ludDsKIH0gIAp9Cn0=
Error: Tracking ID:856af8defafa48859510dc5cd24e60f7-G:0-TimeStamp:07/21/2021 19:58:24

```

## Steps to Reproduce
Provide a detailed set of steps to reproduce the bug.
1. Take the file https://raw.githubusercontent.com/Azure/iotedge/master/edge-modules/api-proxy-module/templates/nginx_default_config.conf and base64 encode it.
2. attempt to set the tag proxy_config to the encoded value

### Output of `iotedge check`

<details>
<summary>Click here</summary>

```

This error happens in the Azure Portal.

```
</details>

### Device Information
* Host OS [e.g. Ubuntu 18.04, Windows Server IoT 2019]:  Ubuntu 18.04
* Architecture [e.g. amd64, arm32, arm64]:  amd64
* Container OS [e.g. Linux containers, Windows containers]: Linux

### Runtime Versions
* aziot-edged [run `iotedge version`]: 
* Edge Agent [image tag (e.g. 1.0.0)]: 1.2.3
* Edge Hub [image tag (e.g. 1.0.0)]: 1.2
* Docker/Moby [run `docker version`]: 20.10.7+azure

Note: when using Windows containers on Windows, run `docker -H npipe:////./pipe/iotedge_moby_engine version` instead

## Logs

<details>
<summary>aziot-edged logs</summary>

```

<Paste here between the triple backticks>

```
</details>

<details>
<summary>edge-agent logs</summary>

```

<Paste here between the triple backticks>

```
</details>

<details>
<summary>edge-hub logs</summary>

```

<Paste here between the triple backticks>

```
</details>

## Additional Information
Please provide any additional information that may be helpful in understanding the issue.

**Comment by ancaantochi** [comment]
There 2 workarounds for this issue:

1. build your own image with the config included: 
                    `git clone https://github.com/Azure/iotedge/tree/master/edge-modules/api-proxy-module` 
                     run `./build.sh -t amd64`
2. get your config to the device and mount it to the container to /app/nginx_default_config.conf
 
      to mount to container from deployment set createOptions:                 

```
 "createOptions": {
     "HostConfig": {
        "Binds":["<HostStoragePath>:<ModuleStoragePath>"]
      }
}
```

**Comment by huguesBouvier** [comment]
You might want to git clone the full iotedge repo: git clone https://github.com/Azure/iotedge. Because API proxy has a dependency on workload API client

**Comment by ancaantochi** [comment]
closing this issue because of no activity, please reopen if needed

**Comment by ksaye** [comment]
@ancaantochi I agree that there are options to work around it, but the documentation of using the twin does not work.

**Comment by deeTEEcee** [comment]
Yeah, it was nice to have a workaround posted but just closing the ticket doesn't solve the issue.

**Comment by vasdee** [comment]
I know this ticket has just been closed, but I feel some options should have been suggested as potential improvements to this issue before closing. 

First potential solution could be for the "main" template to not be overridden but instead _include_ the supplemental configuration provided via the twin.  At least in my case, all I want to do is add an additional directive to the config.

Second solution, split the base64 encoded config into 4k chunks. This is a pattern I currently use to encode >4Kb <32kb files into the twin. The resulting twin in this case would be something like:

```
"proxy_config": {
  "0": "<First chunk of base64 encoded config file>",
  "1": "<Second chunk of base64 encoded config file >",
  ....
  "4": "<Last chunk of base64 encoded config file>"
}
``` 
The module then has an additional step of concatenating the proxy_config object, before base64 decoding.

**Comment by huguesBouvier** [comment]
@deeTEEcee  It looks like the images on dockerhub are gone. They have been moved to https://mcr.microsoft.com/.

**Comment by deeTEEcee** [comment]
Here's some of my findings while testing proxy_config and using docker image tag 1.1.3-linuxamd64:
* Still hit the 4096 tag limit but not on my initial try so it seems inconsistent.
* Upon using `proxy_config` with image tag 1.1.3, the server reports that it picked up the change but it never actually goes through. I had to manually restart the device or docker container in order for it to apply. After that, I ended up failing to remove this from the module twin device config in the interface. At some point, I successfully removed proxy_config away from this module but it took several tries and seems inconsistent. 

I'd say there's potential bugs in both the module being updated and the proxy itself.

**Comment by deeTEEcee** [comment]
> Second solution, split the base64 encoded config into 4k chunks.
🤮  but ill give it a try. still better than me pushing up new images on every change.

**Comment by huguesBouvier** [comment]
I don't think second solution will work unless we update the API proxy code behind.

**Comment by vasdee** [comment]
> I don't think second solution will work unless we update the API proxy code behind.

yes correct. Apologies @deeTEEcee , I was proposing some potential solutions for the IotEdge team to consider. They would need to be changed in the api-proxy code 

I am facing this limitation currently, I would consider making 1 of the changes myself and submitting a PR back, but I would need some buy-in from the iotedge team if they would be happy with that solution before starting

For now, I'll be creating a simple pipeline to use the current api proxy docker image as a base, and simply adding the custom nginx template to the image and building from there. It's not ideal, but it'll work for us

**Comment by deeTEEcee** [comment]
Luckily, it was pretty easy for me to shell into the docker image and then just run `nginx -s reload` everytime i made changes so that'll be my approach.

On an unrelated note, I have to figure out how to utilize workload API on my application module because I need to use the same trusted CA cert that is produced in this module... Unfortunately, we have no documentation for this so I'll have to do some module exploration.

> For now, I'll be creating a simple pipeline to use the current api proxy docker image as a base, and simply adding the custom nginx template to the image and building from there. It's not ideal, but it'll work for us

Makes sense, that's what I did too.
